crates/hyperdrive-math/src/test_utils/agent.rs

```diff
@@ -0,0 +1,455 @@
+use std::{cmp::min, collections::btree_map::Entry};
+
+use ethers::{prelude::EthLogDecode, signers::LocalWallet, types::U256};
+use eyre::Result;
+use fixed_point::FixedPoint;
+use fixed_point_macros::{fixed, uint256};
+use hyperdrive_wrappers::wrappers::ihyperdrive::{Checkpoint, IHyperdriveEvents, Options};
+use rand_chacha::ChaCha8Rng;
+use test_utils::{
+    agent::{Agent, ContractCall, TxOptions},
+    chain::ChainClient,
+};
+use tracing::instrument;
+
+use crate::State;
+
+pub trait HyperdriveMathAgent {
```

Comment:
Does the trait have to be defined first to implement the methods? Could we just implement the trait directly and avoid the duplication?

Reply:
Yup you need it. If you try to do this in hyperdrive-math when Agent is defined in test_utils:
```rust
impl Agent<ChainClient<LocalWallet>, ChaCha8Rng> {
```
You'll get this:
```
cannot define inherent `impl` for a type outside of the crate where the type is defined
define and implement a trait or new type insteadrustc[Click](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B1%5D?1#file%3A%2F%2F%2FUsers%2Fmatthewbrown%2Fsrc%2Fdelv%2Fhyperdrive-rs%2Fcrates%2Fhyperdrive-math%2Fsrc%2Ftest_utils%2Fagent.rs)```